CDAP-3870 Adding logging and metrics support for transforms

gsps1 opened the pull request:
Metrics context is similar to how we record filtered metrics currently. 
For Logging context,  Along with the class name we also add the stage number 
Example : co.cask.cdap.etl.transform.ValidatorTransform.stage-2

In order to show the ValidatorTransform along with the stage, i have modified the LogHandler default pattern to be the set context (``c``) instead of class name (``C``).

JIRA : https://issues.cask.co/browse/CDAP-3870

build running : http://builds.cask.co/browse/CDAP-DUT2966-1

Reopened : https://github.com/caskdata/cdap/pull/4284

chtyim commented on cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/TransformContext.java:
```diff
@@ -55,6 +55,12 @@
+   * Get the stage number of the transform, this is useful for setting the context for logging in transform.
+   * @return {@link Integer} stage number
+   */
+  Integer getStageId();
```
Why returns `Integer` but not `int`? Can it be null?

gsps1 replied:
it cannot, have changed it.

albertshau commented on cdap-watchdog/src/main/java/co/cask/cdap/logging/LoggingConfiguration.java:
```diff
@@ -51,7 +51,7 @@
   public static final String LOG_META_DATA_TABLE = "log.meta";
   // Defaults
-  public static final String DEFAULT_LOG_PATTERN = "%d{ISO8601} - %-5p [%t:%C{1}@%L] - %m%n";
+  public static final String DEFAULT_LOG_PATTERN = "%d{ISO8601} - %-5p [%t:%c{40}@%L] - %m%n";
```
what does this change?

gsps1 replied:
with the pattern ``[%t:%C{1}@%L]`` we will always print the className as the context, so when we emit the log from javascript, the context would be ``sun.reflect.NativeMethodAccessorImpl``, instead we want to have the transform context, that's why we have to change the pattern to be 
[%t:%c{40}@%L] , the 40 is to show the stage number as well as the class name {Eg: validatorTransform}

albertshau replied:
won't this change all our logs? Is there some way to do this only for the logger in the transform?

gsps1 replied:
this is used by the LogHandler endpoint, so this will only be reflected in the UI and REST calls and not the logs we write to the files.

albertshau replied:
so what will the old logs look like now? Seems wrong to change UI and REST output just for this transform.

albertshau replied:
ok understand now, this looks fine

chtyim commented on cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/transform/ScriptFilterTransform.java:
```diff
@@ -98,6 +102,10 @@ private void init() {
 
+    // initialize metrics and logging
+    engine.put("metrics", metrics);
+    engine.put("LOG", logger);
```
Why `metrics` is all small, but `LOG` is all caps?

gsps1 replied:
i tried to be consistent with our usage in java code, but can be changed if this is not appropriate.

chtyim replied:
In java, because `LOG` is a global static while `metrics` is a class field. Since this is Javascript, we can have a convention saying all global variables available to the script is in all caps. Better yet, we can namespace the CDAP variables by a `CDAP` namespace:

```java
  engine.put("CDAP", new ScriptGlobal());
  engine.eval("CDAP.logger.info('Hello, World')");

  public static final class ScriptGlobal {
    public final Logger logger;
    public final Metrics metrics;
   ...
  }
```

gsps1 replied:
ok, i am just wondering how we should set validators.. should we namespace them with CDAP as well ? currently its set as separate global variables and referenced liked, ``Eg: coreValidator.maxLength(input.body, 10)``

chtyim replied:
Is that specific for one transform plugin?

gsps1 replied:
yes

chtyim replied:
In general, there are two ways to do this. One is through a global context object (like `CDAP` I mentioned above). The other is through a context object passed to the function. Personally I like the function argument better since using global is always less preferred.

albertshau commented on cdap-app-templates/cdap-etl/cdap-etl-lib/src/test/java/co/cask/cdap/etl/transform/ScriptTransformTest.java:
```diff
@@ -197,6 +197,8 @@ public void testComplex() throws Exception {
       "  var e = input.inner1.list[0].e;\n" +
       "  var val = power(pi.val, 3) + power(e.val, 2);\n" +
       "  print(pi); print(e);\n" +
+      "  metrics.count(\"script.transform.count\", 1);\n" +
```
can we check that this was emitted somehow? I think there are some tests that use a mock metrics.

albertshau commented on cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/TransformContext.java:
```diff
@@ -55,6 +55,12 @@
+   * Get the stage number of the transform, this is useful for setting the context for logging in transform.
+   * @return stage number
+   */
+  int getStageId();
```
what about returning the PluginID.  Would have more info, including the stage id.

albertshau replied:
ignore this, plugins already know their type and name

albertshau commented on cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/transform/ScriptFilterTransform.java:
```diff
@@ -98,6 +102,10 @@ private void init() {
     engine = manager.getEngineByName("JavaScript");
 
+    // initialize metrics and logging
+    engine.put("metrics", metrics);
```
should describe how to use these in the docs, otherwise nobody will know they exist.

gsps1 commented:
@chtyim @albertshau have addressed your comments. Also added the relevant UI changes (updating json) and docs changes to this PR. @awholegunch can you please review the UI changes. Thanks

albertshau commented:
a couple minor comments, plus a question about the log format change.

awholegunch commented on cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/transform/ValidatorTransform.java:
```diff
@@ -56,21 +57,31 @@
+    "         metrics.count(\"invalid.body.size\", 1);" +
+    "      }" +
+    "      if (!isValid) {" +
+    "       logger.warn(\"Validation failed for record {}\", input);" +
```
Does this work in Javascript? Should it not be something such as:
```
logger.warn(\"Validation failed for record \" + input.id);" +

gsps1 replied:
it will work, as its calling a java object

albertshau commented:
:+1:

gsps1 commented:
Rebased, will merge after build - http://builds.cask.co/browse/CDAP-DUT2966-3

gsps1 commented:
Green, Merging
